internal/cloudfront/service.go

```diff
@@ -0,0 +1,340 @@
+		errs = multierror.Append(errs, s.upsertCDNStatus(ctx, cdnStatus))
+	}
+
+	return s.handleResult(ing, cdnStatus, errs)
+}
+
+func (s *Service) desiredState(ctx context.Context, reconciling k8s.CDNIngress, obj client.Object) ([]k8s.CDNIngress, Distribution, error) {
+	desiredIngresses, err := s.desiredIngresses(ctx, reconciling, obj)
+	if err != nil {
+		return nil, Distribution{}, err
+	}
+
+	sharedParams, err := k8s.NewSharedIngressParams(desiredIngresses)
+	if err != nil {
+		return nil, Distribution{}, fmt.Errorf("shared ingress params: %v", err)
+	}
+
+	existingDistARN, err := s.DistRepo.ARNByGroup(reconciling.Group)
+	if err != nil && !errors.Is(err, ErrDistNotFound) {
+		return nil, Distribution{}, fmt.Errorf("fetching existing CloudFront ID based on group (%s): %v", reconciling.Group, err)
+	}
+
+	desiredDist, err := newDistribution(desiredIngresses, reconciling.Group, sharedParams.WebACLARN, existingDistARN, s.Config)
+	if err != nil {
+		return nil, Distribution{}, fmt.Errorf("building desired distribution: %v", err)
+	}
+
+	return desiredIngresses, desiredDist, nil
+}
+
+func (s *Service) desiredIngresses(ctx context.Context, reconciling k8s.CDNIngress, obj client.Object) ([]k8s.CDNIngress, error) {
+	desiredIngresses, err := s.Fetcher.FetchBy(ctx, s.isPartOfDesiredState(reconciling))
+	if err != nil {
+		return nil, fmt.Errorf("listing active Ingresses that belong to group %s: %v", reconciling.Group, err)
+	}
+	return desiredIngresses, nil
+}
+
+func (s *Service) isPartOfDesiredState(reconciling k8s.CDNIngress) func(k8s.CDNIngress) bool {
+	return func(ing k8s.CDNIngress) bool {
+		isPartOfGroup := ing.Group == reconciling.Group
+		hasBeenProvisioned := len(ing.LoadBalancerHost) > 0
+		return !ing.IsBeingRemoved && isPartOfGroup && hasBeenProvisioned
+	}
+}
+
+func (s *Service) fetchOrGenerateCDNStatus(desiredIngs []k8s.CDNIngress, dist Distribution) (*v1alpha1.CDNStatus, error) {
+	status := &v1alpha1.CDNStatus{}
+	key := client.ObjectKey{Name: dist.Group}
+
+	err := s.Client.Get(context.Background(), key, status)
+	if k8serrors.IsNotFound(err) {
+		return newCDNStatus(desiredIngs, dist), nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("fetching CDNStatus: %v", err)
+	}
+
+	return status, nil
+}
+
+func newCDNStatus(ings []k8s.CDNIngress, dist Distribution) *v1alpha1.CDNStatus {
+	status := &v1alpha1.CDNStatus{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: dist.Group,
+		},
+	}
+
+	for _, ing := range ings {
+		status.SetIngressRef(false, ing)
+	}
+
+	return status
+}
+
+func (s *Service) syncDist(ctx context.Context, desiredDist Distribution, cdnStatus *v1alpha1.CDNStatus, ing client.Object) (Distribution, error) {
+	if desiredDist.IsEmpty() {
+		return desiredDist, s.deleteDistribution(ctx, desiredDist)
+	}
+	return s.upsertDistribution(ctx, desiredDist, cdnStatus, ing)
+}
+
+func (s *Service) upsertDistribution(ctx context.Context, dist Distribution, status *v1alpha1.CDNStatus, ing client.Object) (Distribution, error) {
```

Comment:
For me, it makes sense to export this function as a service to encapsulate the domain and persistence layer. e.g.:

```go
svc := distribution.NewService()
svc.UpsertDistribution(dist)
```
it's strange for me that the service handles objects at the Kubernetes and controller level

Reply:
> it's strange for me that the service handles objects at the Kubernetes and controller level

Yes, this needs to be improved. Ideally, the service would only deal with `k8s.CDNIngress`, but I had to stop refactoring at some point lol

> For me, it makes sense to export this function as a service to encapsulate the domain and persistence layer. e.g.

For now I disagree based on the declarative interface argument I made in the other thread.

Reply:
Ok